cpp/src/arrow/array/builder_binary.h

```diff
@@ -396,52 +401,62 @@ class BaseBinaryBuilder : public ArrayBuilder {
 /// \brief Builder class for variable-length binary data
 class ARROW_EXPORT BinaryBuilder : public BaseBinaryBuilder<BinaryType> {
  public:
-  using BaseBinaryBuilder::BaseBinaryBuilder;
+  explicit BinaryBuilder(MemoryPool* pool = default_memory_pool())
+      : BaseBinaryBuilder(binary(), pool) {}
+
+  BinaryBuilder(const std::shared_ptr<DataType> &type, MemoryPool *pool)
+      : BaseBinaryBuilder(type, pool) {}
```

Comment:
For this constructor and the other constructors below which accept an explicit type, please do one of:
- delete them
- make them protected as well
- rewrite them to `BinaryBuilder(const std::shared_ptr<ExtensionType> &type, MemoryPool *pool = default_memory_pool())`, since the only valid case I can see for allowing public construction of a `BinaryBuilder` with an output type other than `binary()` is building an extension array (which AFAICT is not used anywhere nor is it tested so my preference would be deletion for now)

Reply:
This constructor is used in other places of the code with templates. It seems all the builders are expected to have this constructor.